### PR TITLE
Feature/834 com cob offset

### PR DESCRIPTION
**Pull request by JulianHammerl**
* **Tickets addressed:** MAXGNC-834
* **Review:** By commit
* **Merge strategy:** Merge (no squash)

## Description

The cobConverter module was updated to include the center of mass (COM) offset estimation due to the phase angle correction. In addition to the unit vector pointing to the center of brightness (COB), the module now outputs another message containing the unit vector pointing to the center of mass (COM) as well as an output message containing more information about the COM computation. The noise (covariance matrix) for the COM unit vector is the same as for the COB unit vector.

The module now requires an input parameter to specify which method should be used to compute the phase angle correction ("NoCorrection" needs to be chosen to not apply any correction).

## Verification

The UnitTest was updated accordingly.

## Documentation

The documentation was updated accordingly.

## Future work

N/A

### CODE REVIEW

File: src/architecture/msgPayloadDefCpp/OpNavCOMMsgPayload.h [hunk 1]
```diff
@@ -0,0 +1,44 @@
+/*
+ ISC License
+
+ Copyright (c) 2016, Autonomous Vehicle Systems Lab, University of Colorado at Boulder
```
Comment (patkenneally) [inline src/architecture/msgPayloadDefCpp/OpNavCOMMsgPayload.h]: :small_red_triangle_down: Let's change this line to 

`Copyright (c) 2024, Laboratory for Atmospheric and Space Physics, University of Colorado at Boulder`

File: src/architecture/msgPayloadDefCpp/OpNavCOMMsgPayload.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+ */
+
+#ifndef COMOPNAVMSG_H
+#define COMOPNAVMSG_H
+
+#include <array>
```
Comment (patkenneally) [inline src/architecture/msgPayloadDefCpp/OpNavCOMMsgPayload.h]: :small_red_triangle_down: Remove unused.

File: src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp [hunk 3]
```diff
@@ -143,5 +143,19 @@ void CobConverter::UpdateState(uint64_t CurrentSimNanos)
     }
 
     this->opnavUnitVecCOBOutMsg.write(&uVecCOBMsgBuffer, this->moduleID, CurrentSimNanos);
+}
+
+/*! Set the object radius
+    @param double radiusInput [m]
+    @return void
+    */
+void CobConverter::setRadius(const double radiusInput){
```
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp]: :large_orange_diamond: The parameter name can just be `radius`.

File: src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp [hunk 4]
```diff
@@ -96,25 +96,75 @@ void CobConverter::UpdateState(uint64_t CurrentSimNanos)
...
+        Eigen::Vector3d shat_C;
+        shat_C = dcm_CB * shat_B;
+        double phi = atan2(shat_C[1], shat_C[0]); // sun direction in image plane
+
+        double focalLength = cameraSpecs.ppFocalLength; // focal length in m
+        double Kx = dX / focalLength; // camera parameter from Bhaskaran paper
+        double Rc = this->objectRadius * Kx * focalLength / rho; // object radius in pixels
+
+        double gamma = 0; // offset factor between Center of Mass and Center of Brightness
+        bool validCOM;
```
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp]: :large_orange_diamond: If you initialize `validCom` to False then you can remove your final else condition.

File: src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp [hunk 5]
```diff
@@ -96,25 +96,75 @@ void CobConverter::UpdateState(uint64_t CurrentSimNanos)
                                           0., 1./dY, -vp/dY,
                                           0., 0., 1.;
 
+        /*! - phase angle correction */
+        Eigen::Vector3d rhat_N;
```
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp]: :large_orange_diamond: I realize your following the convention in the module, but I think we should refactor/move to declaring/assigning on the same line when we can. E.g. 
`Eigen::Vector3d rhat_N = cArray2EigenVector3d(ephemBuffer.r_BdyZero_N).normalized();`
or
`double rho = cArray2EigenVector3d(ephemBuffer.r_BdyZero_N).norm();`

File: src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp [hunk 6]
```diff
@@ -51,42 +51,61 @@ void CobConverter::UpdateState(uint64_t CurrentSimNanos)
 {
     CameraConfigMsgPayload cameraSpecs = this->cameraConfigInMsg();
     OpNavCOBMsgPayload cobMsgBuffer = this->opnavCOBInMsg();
-    OpNavUnitVecMsgPayload uVecMsgBuffer = this->opnavUnitVecOutMsg.zeroMsgPayload;
```
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp:54]: You must know what I'm going to say about this commit ;)

File: src/fswAlgorithms/opticalNavigation/cobConverter/_UnitTest/test_cobConverter.py [hunk 7]
```diff
@@ -28,141 +28,243 @@
...
+    valid_COM = dataLogCOM.valid[0]
+    rhat_COM_N = dataLogUnitVecCOM.rhat_BN_N[0]
 
-    for i in range(len(outputCovar[-1, 0:])):
-        if np.abs((covar_Nexp[i] - outputCovar[-1, i])) > tollerance and np.isnan(outputTime.any()):
-            testFailCount += 1
-            testMessages.append("Covar Check in cobConvert")
+    # make sure module output data is correct
+    tolerance = 1e-10
+    np.testing.assert_allclose(rhat_COB_N,
```
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/cobConverter/_UnitTest/test_cobConverter.py:225]: Thank you!

**Review by patkenneally — APPROVED** [review]
:small_red_triangle_down: *Issues* to address before merge
:large_orange_diamond: *Requests* that should not block merge, but should at least be discussed
:large_blue_circle: *Recommendations* that can be ignored if desired

File: src/architecture/msgPayloadDefCpp/OpNavCOMMsgPayload.h (hunk 1 above)
Comment (JulianHammerl) [reply src/architecture/msgPayloadDefCpp/OpNavCOMMsgPayload.h]: Done.

File: src/architecture/msgPayloadDefCpp/OpNavCOMMsgPayload.h (hunk 2 above)
Comment (JulianHammerl) [reply src/architecture/msgPayloadDefCpp/OpNavCOMMsgPayload.h]: Done.

File: src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp (hunk 3 above)
Comment (JulianHammerl) [reply src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp]: Done.

File: src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp (hunk 5 above)
Comment (JulianHammerl) [reply src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp]: Done. I also made these changes to the first commit (refactor of existing module).

File: src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp (hunk 4 above)
Comment (JulianHammerl) [reply src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp]: Done.

File: src/fswAlgorithms/opticalNavigation/cobConverter/_UnitTest/test_cobConverter.py (hunk 7 above)
Comment (JulianHammerl) [reply src/fswAlgorithms/opticalNavigation/cobConverter/_UnitTest/test_cobConverter.py:225]: 🫶

File: src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp (hunk 6 above)
Comment (JulianHammerl) [reply src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp:54]: Split the commit up into several commits.

**Pull request merged**